### PR TITLE
Prototype initial deploy scripting API

### PR DESCRIPTION
**Pull request by mkaput**
cc #998

---

**Stack**:
- #1035
- #1034 ⬅
- #1017
- #1016

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

**Comment by mkaput** [comment]
question: should I include a series of tests showcasing what I see here? These tests would assert that `NotImplemented` is raised

**Review by Arcticae — COMMENTED** [review]
Overall - LGTM, some thoughts:
As for the tests - i don't think there's a need to have them
As for the implementation stubs - are those the same we already have in migrations? If so, shouldn't we just be using those implementations?

**Review by Arcticae — COMMENTED** [review]
Overall - LGTM, some thoughts:
As for the tests - i don't think there's a need to have them
As for the implementation stubs - are those the same we already have in migrations? If so, shouldn't we just be using those implementations?

**Review by Arcticae — COMMENTED** [review]
Overall - LGTM, some thoughts:
As for the tests - i don't think there's a need to have them
As for the implementation stubs - are those the same we already have in migrations? If so, shouldn't we just be using those implementations?

**Comment by kasperski95** [comment]
I am not sure if reusing migration cheatcode signatures is a good idea due. A user needs to import every function.

At the other, we can try mimicking CLI interface so that the user can reuse that knowledge.
If you know how the CLI looks like, you know how to use configuration file (more or less) and you'll know how to use scripts. CLI could be treated as a source of truth. Or `protostar.scripting` should be a source of truth for `protostar.commands`. Maybe `protostar.scripting` should be renamed to `protostar.use_cases`?

> question: should I include a series of tests showcasing what I see here? These tests would assert that NotImplemented is raised

We don't have to be that strict.

**Comment by mkaput** [comment]
> A user needs to import every function.
> 
> At the other, we can try mimicking CLI interface so that the user can reuse that knowledge. If you know how the CLI looks like, you know how to use configuration file (more or less) and you'll know how to use scripts. CLI could be treated as a source of truth. Or `protostar.scripting` should be a source of truth for `protostar.commands`. Maybe `protostar.scripting` should be renamed to `protostar.use_cases`?

@kasperski95 my idea for this API in long term is:

Non-transactional code:
```python
from protostar.scripting.deployments import *

call(something)
invoke(something)
declare(something)
# I don't know how to model deploy yet
```

Transactional code:
```python
from protostar.scripting.deployments import *

with multicall(...) as mc:
    mc.call(something)
    mc.invoke(something)
    mc.declare(something)
    # I don't know how to model deploy yet
```

File: protostar/scripting/deployments/declare.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+@dataclass
+class DeclaredContract:
+    class_hash: int
+
+
+def declare(
+    contract: str,
+    *,
+    wait_for_acceptance: Optional[bool] = None,
```
Comment (mkaput) [inline protostar/scripting/deployments/declare.py:15]: I wonder if we shouldn't make `wait_for_acceptance` be `True` by default, and non-optional, i.e. use `wait_for_acceptance: bool = True` everywhere

File: protostar/scripting/typing.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+import protostar.starknet_gateway.gateway_facade as _gf
+
+PythonData = _dt.PythonData
+CairoOrPythonData = _dt.CairoOrPythonData
+FromPythonTransformer = _dt.FromPythonTransformer
+ToPythonTransformer = _dt.ToPythonTransformer
+
+ContractFunctionInputType = _gf.ContractFunctionInputType
+Wei = _gf.Wei
+Fee = _gf.Fee
```
Comment (mkaput) [inline protostar/scripting/typing.py:11]: I wonder if we couldn't re-export these from `protostar.scripting` directly 🤔

**Comment by MaksymilianDemitraszek** [comment]
```
with multicall(...) as mc:
    mc.call(something)
    mc.invoke(something)
    mc.declare(something)
```
I think you can multicall only invokes, but contextmanager API is fine

File: protostar/scripting/deployments/invoke.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from typing import Optional
+
+from protostar.scripting.typing import CairoOrPythonData, Fee
+
+
+def invoke(
```
Comment (MaksymilianDemitraszek) [inline protostar/scripting/deployments/invoke.py:6]: I wonder if we shouldn't have separate function for deploy to make it simpler to user, even though it is invoke under the hood

File: protostar/scripting/deployments/declare.py (hunk 1 above)
Comment (kasperski95) [reply protostar/scripting/deployments/declare.py:15]: Maybe `wait_for_acceptance` should be a generic function.

File: protostar/scripting/deployments/invoke.py (hunk 3 above)
Comment (mkaput) [reply protostar/scripting/deployments/invoke.py:6]: Yes, I agree. But I don't know what signature would make sense yet and is more kind of addition, not necessity initially.

File: protostar/scripting/deployments/declare.py (hunk 1 above)
Comment (mkaput) [reply protostar/scripting/deployments/declare.py:15]: > Maybe `wait_for_acceptance` should be a generic function.

I don't understand, could you elaborate please?

Comment (kasperski95) [reply protostar/scripting/deployments/declare.py:15]:
```py
declare_tx = declare("main")
awaited_declare_tx = wait_for_acceptance(declare_tx)
assert isinstance(awaited_declare_tx, DeclareTx)
```

Comment (mkaput) [reply protostar/scripting/deployments/declare.py:15]: Isn't this too noisy?

Also this sounds like futures, maybe we should reuse that ones?

Comment (mkaput) [reply protostar/scripting/deployments/declare.py:15]: FYI I can allow top-level awaits in these scripts

**Comment by mkaput** [comment]
new PR: #1059

**Pull request merged**